### ISSUE TITLE
simple_seq2seq.py teacher forcing training time

### ISSUE THREAD

**Issue by maxdel** [allenai/allennlp#602]
If scheduled_sampling_ratio is set to 1 (no teacher forcing), it takes ~3 hours per epoch on my GPU.
But if scheduled_sampling_ratio is set to 0 (teacher forcing), it takes ~8 hours per epoch on my GPU.

Any thoughts on that?

**Comment by matt-gardner** [comment]
That's surprising to me.  The only place in the code that makes a difference is here: https://github.com/allenai/allennlp/blob/01291ee822687667ec0557c5c358dde02b7c89d8/allennlp/models/encoder_decoders/simple_seq2seq.py#L146-L155

This suggests that the slicing operation (line 147) is way more expensive than I thought.  What happens if you do something like [`targets.split(split_size=1, dim=1)`](http://pytorch.org/docs/master/torch.html#torch.split) above the loop to pre-slice everything, and just index into that split list?  Assuming `torch.split()` is implemented efficiently, this would be a single GPU operation instead of `num_timesteps` slice operations, which could be significantly faster.  Looking at the implementation of `torch.split()`, I'm not certain this will save any time, but it _might_, so it seems like it's worth a try.

**Comment by maxdel** [comment]
That's surprising to me as well.
I will try your suggestion soon.

**Comment by matt-gardner** [comment]
@brendan-ai2, did you experiment with any kind of teacher forcing in your event2mind model?  Just curious if you've seen anything related to this.  I'm probably going to close this issue as stale, as none of us are working on the seq2seq model at this point.

**Comment by maxdel** [comment]
@matt-gardner Did not have time to try it at that point.
I think the issue can be closed now and I will reopen it if I have results.

**Comment by maxdel** [comment]
Just made a test and the epoch duration in both cases is about the same:

No teacher forcing:
> 2018-09-15 03:59:55,854 - INFO - allennlp.training.trainer - Epoch duration: 00:07:54
> 2018-09-15 03:59:55,854 - INFO - allennlp.training.trainer - Estimated training time remaining: 2 days, 18:29:10

Teacher forcing:

> 2018-09-15 19:11:42,616 - INFO - allennlp.training.trainer - Epoch duration: 00:08:09
> 2018-09-15 19:11:42,616 - INFO - allennlp.training.trainer - Estimated training time remaining: 2 days, 19:54:02

 Probably the problem was unrelated to the codebase.

**Comment by matt-gardner** [comment]
Thanks for checking!
